crates/rust-analyzer/src/handlers.rs

```diff
@@ -568,15 +571,86 @@ pub(crate) fn handle_completion(
     };
     let line_index = snap.analysis.file_line_index(position.file_id)?;
     let line_endings = snap.file_line_endings(position.file_id);
+    let mut completion_resolve_data = FxHashMap::default();
+
     let items: Vec<CompletionItem> = items
         .into_iter()
-        .flat_map(|item| to_proto::completion_item(&line_index, line_endings, item))
+        .enumerate()
+        .flat_map(|(item_index, item)| {
+            let mut new_completion_items =
+                to_proto::completion_item(&line_index, line_endings, item.clone());
+
+            if snap.config.completion.resolve_additional_edits_lazily() {
+                if let Some(import_edit) = item.import_to_add() {
+                    completion_resolve_data.insert(
+                        item_index,
+                        CompletionResolveData {
+                            file_id: position.file_id,
+                            import_edit: import_edit.clone(),
+                        },
+                    );
+
+                    let item_id = serde_json::to_value(&item_index)
+                        .expect(&format!("Should be able to serialize usize value {}", item_index));
+                    for new_item in &mut new_completion_items {
+                        new_item.data = Some(item_id.clone());
+                    }
+                }
+            }
+
+            new_completion_items
+        })
         .collect();
 
+    global_state.completion_resolve_data = completion_resolve_data;
```

Comment:
As my previous comment suggested: we shouldn't need to do this. It's ok to repeat some work during resolve.